Feature/fusesoc scoped constraints

rbrglez opened the pull request:
#160

Added scoped constraints to fusesoc development cores (base, intf).

I have reclassified critical warning to warning (check out olo_base_constraints_amd.tcl and olo_intf_constraints_amd.tcl).
I have tested vivado_tutorial fusesoc core with vivado 2023.2 and the scoped constraints were used successfully.
I have tested quartus_tutorial fusesoc core with quartus standard 24.1 and the scoped constraints weren't included (as expected). @obruendl could you test if other tutorials work as expected?

TODOs: 
- Check vivado_tutorial scripted flow (import_sources.tcl).
- Update official base and intf fusesoc cores

obruendl commented on .gitignore:
This depends on where you set-up your fusesoc. You can easily set-up fusesoc in another directory.

I'd like to not add lines to .gitignore that are anyhow related to the project structure.

rbrglez replied:
this is unrelated to folder structure. you can set-up fusesoc in open-logic root:
```
fusesoc library add src
fusesoc library add doc
```

or in e.g. src folder:

```
fusesoc library add .
fusesoc library add ../doc
```

gitignore fusesoc additions will ignore fusesoc.conf and build/ in both examples

obruendl commented on src/base/tcl/olo_base_constraints_amd.tcl:
Are you sure this works as intended?

I believe the "read_xdc" command is executed immediately and not "LATE" as required. I remember I had quite some troubles aroudn this topic - which is why I ended up creating a intermediate file with order "LATE" whtat does execute the read_xdc commands.

rbrglez replied:
It looks like the LATE command is doing something. When I have removed the LATE command I got Critical warnings.

obruendl commented on src/intf/tcl/olo_intf_constraints_amd.tcl:
Same comment as for the _base_ file

rbrglez commented on tools/vivado/all_constraints_amd.tcl:
I have updated this script, which is used by scripted_build.tcl in Vivado Tutorial. I have tested it and it works fine.

rbrglez commented on src/base/tcl/constraints_amd.tcl:
I renamed the file to olo_base_constraints_amd.tcl.

Previously, I copied all TCL files from the intf and base folders directly into the FuseSoC-generated workspace root. Now, I copy them into their respective base/ and intf/ subfolders, so renaming is no longer strictly necessary.

However, I prefer to keep the renamed files. I think it’s better to maintain a consistent naming convention by appending olo_base or olo_intf to the constraints_amd.tcl files.

I have also updated scripted_build.tcl to work correctly with these renamed constraint files.

rbrglez commented on tools/vivado/import_sources.tcl:
I've simplified import_sources.tcl, and I prefer this new approach. Previously, when using scripted_build.tcl, Vivado displayed the constraints like this:

![image](https://github.com/user-attachments/assets/02b5f29b-4e11-48e4-b389-087b066c018a)

In that setup, you could only see that the constraints were sourcing another file, without any indication of which specific constraints were actually being applied.

With the updated version, the constraints now appear like this:

![image](https://github.com/user-attachments/assets/6909729f-485a-4285-9a59-37cf69eff886)

This is much more descriptive. Vivado now shows the actual constraint files being used.

obruendl commented:
I'll try out the changes. I want to take myself the time to test it in detail and also check if there are no impacts for version controlling the resulting vivado project (no absolute oaths in the xpr).

I remember that I had a tough time to find a working solution. I'd be happy to incorporate your simpler approach if it really has no significant drawbacks.

Btw: I am on holiday in the coming days, so testing it could take a week or so

obruendl commented:
I changed the target branch to "develop" - there I integrate features. Merges to main only happen on release.

obruendl reviewed (approved):
Looks OK now

obruendl commented:
@rbrglez 

I checked today, looks all good. I also updated the Python .core generator and tested the result - it seems to work fine but it'd be great if you could check if you agree that the generated .core files match yours content wise.

I tend to keep git History clean, so I usually do squash-merge. Unfortnuately this ends up as a single commit with my name on it. However, you would like to preserve your name on the commit - you deserve the credits and it's also good for open logic to show that people are contributing. Hence **Could you rebase/squash all changes into a single commit**? 

This would allow to have a clean Git history (one commit per feature added) but it preserves your name on this commit.

obruendl commented:
@rbrglez 

When you cleanup the history of this PR (rebase everything into one commit as asked before), please make sure the commit has the correct e-mail address and name in (the one you use for your GitHub account). I just set up the CLA and an automatic check for all contributors to sign it. For your commits to be accepted, the commits of course must be traceable to your github account.

rbrglez commented:
@obruendl 

I have signed CLA and Squashed all commits in one "FEATURE" commit.

I modified `core.template` by adding quotes around certain values. This change aligns with recommendations in the FuseSoC documentation: https://fusesoc.readthedocs.io/en/stable/user/build_system/core_files.html.

I performed basic tests using FuseSoC by building both the `vivado_tutorial` and `quartus_tutorial` examples. Everything appeared to work correctly with the new changes.
